sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala

```diff
@@ -108,18 +108,22 @@ object IntervalUtils {
 
   def castStringToYMInterval(
       input: UTF8String,
-      // TODO(SPARK-35768): Take into account year-month interval fields in cast
       startField: Byte,
       endField: Byte): Int = {
+
+    def truncatedMonth(month: String) : String = {
+      if (endField == YearMonthIntervalType.YEAR) "0" else month
+    }
+
     input.trimAll().toString match {
-      case yearMonthRegex("-", year, month) => toYMInterval(year, month, -1)
-      case yearMonthRegex(_, year, month) => toYMInterval(year, month, 1)
+      case yearMonthRegex("-", year, month) => toYMInterval(year, truncatedMonth(month), -1)
+      case yearMonthRegex(_, year, month) => toYMInterval(year, truncatedMonth(month), 1)
       case yearMonthLiteralRegex(firstSign, secondSign, year, month) =>
         (firstSign, secondSign) match {
-          case ("-", "-") => toYMInterval(year, month, 1)
-          case ("-", _) => toYMInterval(year, month, -1)
-          case (_, "-") => toYMInterval(year, month, -1)
-          case (_, _) => toYMInterval(year, month, 1)
+          case ("-", "-") => toYMInterval(year, truncatedMonth(month), 1)
+          case ("-", _) => toYMInterval(year, truncatedMonth(month), -1)
+          case (_, "-") => toYMInterval(year, truncatedMonth(month), -1)
+          case (_, _) => toYMInterval(year, truncatedMonth(month), 1)
         }
       case _ => throw new IllegalArgumentException(
```

Comment:
Could you change the error message, please. It should reflect `startField` and `endField`.

Reply:
> Could you change the error message, please. It should reflect `startField` and `endField`.

How about current?